pina/geometry/union_domain.py

```diff
@@ -108,11 +109,19 @@ def sample(self, n, mode='random', variables='all'):
         num_points = n // len(self.geometries)
 
         # sample the points
-        for i, geometry in enumerate(self.geometries):
+        # NB. geometries as shuffled since if we sample
+        # multiple times just one point, we would end
+        # up sampling only from the first geometry.
+        iter_ = random.sample(self.geometries, len(self.geometries))
+        for i, geometry in enumerate(iter_):
             # add to sample total if remainder is not 0
             if i < remainder:
                 num_points += 1
+            # sampling
             sampled_points.append(geometry.sample(num_points, mode, variables))
+            # in case number of sampled points is smaller than the number of geometries
+            if len(sampled_points) >= n:
+                break
```

Comment:
What is the meaning of those lines? I got the shuffle of the geometries, but not the remainder to be honest.
Are we sure that in this case the number of sampled points is the on provided by the user?

Reply:
Yes, basically it could happen that the user specifies a number which is not an exact multiple of the number of geometries. If we do only the integer division we then undersample, so we need the remainder.

For example, suppose we have `4` geometries and ask for `22` samples. Then `22//4 = 5` so each geometry sample four points. But like so we miss 2 points, which is exactly the remainder. Now, by definition the remainder is always smaller than the number of geometries, so with `i<remainder` we only add one more point to sample in the specific geometry.

While responding to you I notice a bug, because we are updating `num_points` which is the same for all geometries. The correct way to do it is by removing line 118-119, and write simply 

``` sampled_points.append(geometry.sample(num_points + int(i < remainder), mode, variables))```

I will update it now.

Reply:
In [9c89a11](https://github.com/mathLab/PINA/pull/118/commits/9c89a11f2a96a2e0335eab298cac9cec0b2575f6) I fixed the bug you reported, I added the test in `test_union.py` to check the number of points in two different cases:
1. The number of points is divisible by the number of geometries
2. The number of points is not divisible by the number of geometries

I also made modification in `check_consistency` after merging #115, so now the branch does not have any conflict and the tests run smoothly.